Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php

```diff
@@ -378,40 +399,112 @@ protected function resolveRoutePathForNode(NodeInterface $node)
-                    $relativeNodePathSegments[] = $node->getName();
-                    $foundNodeInThisSegment = true;
-                    break;
+        // The DBAL implementation below does not support multiple dimensions yet,
+        // therefore fall back to implementation using the CR API if needed.
+        $dimensionPresets = $this->contentDimensionPresetSource->getAllPresets();
+        if (count($dimensionPresets) > 1 || (count($dimensionPresets) === 1 && count(current($dimensionPresets)['presets']) > 1)) {
+            $node = $siteNode;
+            $relativeNodePathSegments = [];
+
+            foreach (explode('/', $relativeRequestPath) as $pathSegment) {
+                $foundNodeInThisSegment = false;
+                foreach ($node->getChildNodes('Neos.Neos:Document') as $node) {
+                    /** @var NodeInterface $node */
+                    if ($node->getProperty('uriPathSegment') === $pathSegment) {
+                        $relativeNodePathSegments[] = $node->getName();
+                        $foundNodeInThisSegment = true;
+                        break;
+                    }
+                }
+                if (!$foundNodeInThisSegment) {
+                    return '';
                 }
             }
-            if (!$foundNodeInThisSegment) {
-                return false;
-            }
+        } else {
+            $relativeNodePathSegments = $this->findRelativeNodePathByUriPathSegmentPropertiesUsingSql($siteNode, $relativeRequestPath);
         }
 
         return implode('/', $relativeNodePathSegments);
     }
 
+    /**
+     * Fix for performance issue: if a given node contains thousands of child nodes, this implementation would load all
+     * these nodes into memory in order to compare the current path segment with $node->getProperty('uriPathSegment').
+     *
+     * @param NodeInterface $siteNode
+     * @param string $relativeRequestPath
+     * @return array
+     * @throws \Doctrine\DBAL\DBALException
+     */
+    protected function findRelativeNodePathByUriPathSegmentPropertiesUsingSql(NodeInterface $siteNode, string $relativeRequestPath): array
+    {
+        $relativeNodePathSegments = [];
+        $currentNodeRecord = [
+            'path' => $siteNode->getPath()
+        ];
+        $workspaceNames = [$siteNode->getContext()->getWorkspace()->getName()];
+        $baseWorkspaces = $siteNode->getContext()->getWorkspace()->getBaseWorkspaces();
+        array_walk($baseWorkspaces, function (Workspace $workspace) use (&$workspaceNames) {
+            $workspaceNames[] = $workspace->getName();
+        });
+
+        $queryTemplate = 'SELECT path, parentPath
+              FROM neos_contentrepository_domain_model_nodedata AS nodedata
+              WHERE LOWER(CAST(nodedata.properties AS CHAR)) LIKE ?
+              AND nodedata.parentpathhash = ?
+              AND workspace IN (?)
+              ORDER BY FIELD(workspace, ?) ASC
+              LIMIT 1';
```

Comment:
Changing the lines 457-463 to:
```
        $queryTemplate = 'SELECT path, parentPath
              FROM neos_contentrepository_domain_model_nodedata AS nodedata
              WHERE LOWER(CAST(nodedata.properties AS CHAR)) LIKE ?';
        if (!$siteNode->getContext()->isInvisibleContentShown()) {
            $queryTemplate .= '
                AND nodedata.hidden = 0
                AND nodedata.hidden = 0
                AND (nodedata.hiddenbeforedatetime IS NULL OR nodedata.hiddenbeforedatetime > NOW())
                AND (nodedata.hiddenafterdatetime IS NULL OR nodedata.hiddenafterdatetime < NOW())';
        }
        $queryTemplate .= '
              AND nodedata.parentpathhash = ?
              AND workspace IN (?)
              ORDER BY FIELD(workspace, ?) ASC
              LIMIT 1';
```
could solve the "hidden nodes are not filtered" issue

Reply:
Shouldn't the constraint that is indexed (parent path hash?) come first in order to help the query optimizer? I recall something like that. Could be outdated wisdom though

Reply:
yes, might be, good point

Reply:
=>
```
        $queryTemplate = 'SELECT path, parentPath
              FROM neos_contentrepository_domain_model_nodedata AS nodedata
              WHERE LOWER(CAST(nodedata.properties AS CHAR)) LIKE ?
              AND nodedata.parentpathhash = ?';
        if (!$siteNode->getContext()->isInvisibleContentShown()) {
            $queryTemplate .= '
                AND nodedata.removed = 0
                AND nodedata.hidden = 0
                AND (nodedata.hiddenbeforedatetime IS NULL OR nodedata.hiddenbeforedatetime > NOW())
                AND (nodedata.hiddenafterdatetime IS NULL OR nodedata.hiddenafterdatetime < NOW())';
        }
        $queryTemplate .= '
              AND workspace IN (?)
              ORDER BY FIELD(workspace, ?) ASC
              LIMIT 1';
```

FYI: Testing this change currently in a large project

Reply:
Can you verify the query execution plan with the change against your project database, i.e. `EXPLAIN SELECT ..` with different positions of the `parentpathhash` condition?